xarray/plot/plot.py

```diff
@@ -449,10 +449,29 @@ def newplotfunc(darray, x=None, y=None, figsize=None, size=None,
         if imshow_rgb:
             # Don't add a colorbar when showing an image with explicit colors
             add_colorbar = False
+            # Calculate vmin and vmax automatically for `robust=True`
+            if robust:
+                if vmin is None:
+                    vmin = np.nanpercentile(darray, ROBUST_PERCENTILE)
+                if vmax is None:
+                    vmax = np.nanpercentile(darray, 100 - ROBUST_PERCENTILE)
+                robust = False
+            # Scale interval [vmin .. vmax] to [0 .. 1] and clip to bounds
+            if vmin is not None or vmax is not None:
+                vmin = vmin if vmin is not None else darray.min()
+                vmax = vmax if vmax is not None else darray.max()
+                darray = darray.astype('f4' if vmax - vmin < 2 ** 32 else 'f8')
```

Comment:
I think you can use `np.finfo('f4').max` here instead of `2**32`.

Reply:
I think it was uint32 that gave us trouble on Matplotlib.  i.e. way less than f4.max (I think?).  See https://github.com/matplotlib/matplotlib/pull/10133.  I was too lazy to figure out the actual limit (I guess its 2^N, where N is how many bits a float has dynamic range over.  I think its 32-1-8 = 2**23?

Reply:
So you can use `np.iinfo` instead:

```In [12]: np.iinfo(np.uint32)
Out[12]: iinfo(min=0, max=4294967295, dtype=uint32)
```

Reply:
Yeah, but thats not the problem.  The problem is that if you convert (a-vmin)/(1e9) then you can't tell the difference between a=1 and a=2 anymore if you only use float32.  

This came up in Matplotlib because the conversion to float happened with an automatic vmin and vmax, and the user had a large value as a bad-data flag.  *Then* the user subsequently set vmin - vmax = 10 or so, but the convert to float32 had erased all the dynamic between 0 and 10.  

Maybe `xarray` doesn't have this problem because you trust the user to set `vmin` and `vmax` during the `imshow` call.  But if someone grabs the image handle and tries to set the limits after the call they could have this problem.

Reply:
The more deeply I analyse matplotlib/matplotlib#10133, the more suspicious I get - there are actually *two* places you could get loss of precision: subtracting `vmin` (if <<0, loss of precision for high values), and dividing by `(vmin - vmax)`  (loss if vmax is large).

I think for Xarray, the best option is actually just to use `np.float64` in all cases.  We already have a habit of upcasting all the time anyway\*, and if that's lossy you're in trouble anyway.

\* (including `uint16` to `float64` when scaling, which is silly - issue incoming for that)

Reply:
vmin and A are still ints when vmin is removed from A.

Reply:
In which case you are vulnerable to underflow, and hence wraparound of values - that's why I cast to float64 here *first*!

Reply:
Not sure I follow.    

The case we were running into was 
```python
A = np.array( [[1, 2, 1e9], [3, 4, 5]], dtype=np.uint32)
im = plt.imshow(A)
im.set_clim([0., 5.])
```

When the data in the image is first input, `vmin = 1` and `vmax = 1e9`, and the underflow occurred when we cast to `np.float32`, which we did to help accommodate very large images. Hence the check on `vmax - vmin` to see if it is larger than 1e8 and to use `np.float64` for the case instead.  

Casting to `float64` first is great but apparently people sometimes have huge images, hence the extra rigamarole w/ using `float32` if possible.

Reply:
> vmin and A are still ints when vmin is removed from A.

Going and reading the code again, it's not - a_min is subtracted from A_scaled, which has a floating dtype.  In which case you're also safe from overflow/underflow, and have just added handling for loss of precision 😄

I remain happy with this solution - for Xarray, including passing non-RGB images directly to matplotlib - and suggest we take any further discussion to the upstream pull.